diff --git a/src/arm/am335x-boneblack.dts b/src/arm/am335x-boneblack.dts
index 2788ab04..ac761f0b 100644
--- a/src/arm/am335x-boneblack.dts
+++ b/src/arm/am335x-boneblack.dts
@@ -7,7 +7,7 @@
 #include "am33xx.dtsi"
 #include "am335x-bone-common.dtsi"
 #include "am335x-boneblack-common.dtsi"
-#include "am335x-boneblack-hdmi.dtsi"
+#include "BYTD.dtsi"
 
 / {
 	model = "TI AM335x BeagleBone Black";
diff --git a/src/arm/am335x-bonegreen.dts b/src/arm/am335x-bonegreen.dts
index 62ca9c87..259501d1 100644
--- a/src/arm/am335x-bonegreen.dts
+++ b/src/arm/am335x-bonegreen.dts
@@ -7,6 +7,7 @@
 #include "am33xx.dtsi"
 #include "am335x-bone-common.dtsi"
 #include "am335x-bonegreen-common.dtsi"
+#include "BYTD.dtsi"
 
 / {
 	model = "TI AM335x BeagleBone Green";
